peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+    class GameSeries(GameCollection):
+        name: str
+        games: ReadOnly[list[Game]]  # ok: list[Game] is assignable to Collection[Game]
+
+* Nominal subclasses of protocols and ABCs should redeclare read-only attributes
+  in order to implement them, unless the base class initializes them in some way::
+
+    class MyBase(abc.ABC):
+        foo: ReadOnly[int]
+        bar: ReadOnly[str] = "abc"
+        baz: ReadOnly[float]
+
+        def __init__(self, baz: float) -> None:
+            self.baz = baz
+
+        @abstractmethod
+        def pprint(self) -> None: ...
+
+
+    @final
+    class MySubclass(MyBase):
+        # error: MySubclass does not override "foo"
+
+        def pprint(self) -> None:
+            print(self.foo, self.bar, self.baz)
+
+* In a protocol attribute declaration, ``name: ReadOnly[T]`` indicates that a structural
+  subtype must support ``.name`` access, and the returned value is compatible with ``T``::
+
+    class HasName(Protocol):
+        name: ReadOnly[str]
+
+
+    class NamedAttr:
+        name: str
+
+    class NamedProp:
+        @property
+        def name(self) -> str: ...
+
+    class NamedClassVar:
+        name: ClassVar[str]
+
+    class NamedDescriptor:
+        @cached_property
+        def name(self) -> str: ...
+
+    # all of the following are ok
+    has_name: HasName
+    has_name = NamedAttr()
+    has_name = NamedProp()
+    has_name = NamedClassVar
+    has_name = NamedClassVar()
+    has_name = NamedDescriptor()
+
+Interaction with other special types
+------------------------------------
+
+``ReadOnly`` can be used with ``ClassVar`` and ``Annotated`` in any nesting order:
+
+.. code-block:: python
+
+    class Foo:
+        foo: ClassVar[ReadOnly[str]] = "foo"
+        bar: Annotated[ReadOnly[int], Gt(0)]
+
+.. code-block:: python
+
+    class Foo:
+        foo: ReadOnly[ClassVar[str]] = "foo"
+        bar: ReadOnly[Annotated[int, Gt(0)]]
+
+This is consistent with the interaction of ``ReadOnly`` and :class:`typing.TypedDict`
+defined in :pep:`705`.
+
+``ClassVar`` excludes read-only attributes from being assignable to within
+initialization methods.
+
+Rules of ``Final`` should take priority when combined with ``ReadOnly``. As such,
+type checkers may warn on the redundancy of combining the two type qualifiers.
+
+
+Backwards Compatibility
+=======================
+
+This PEP introduces new contexts where ``ReadOnly`` is valid. Programs inspecting
+those places will have to change to support it. This is expected to mainly affect type checkers.
+
+However, caution is advised while using the backported ``typing_extensions.ReadOnly``
+in older versions of Python. Mechanisms inspecting annotations may behave incorrectly
+when encountering ``ReadOnly``; in particular, the ``@dataclass`` decorator
+which `looks for <https://docs.python.org/3/library/dataclasses.html#class-variables>`_
+``ClassVar`` will incorrectly treat ``ReadOnly[ClassVar[...]]`` as an instance attribute.
```

Comment:
Add a recommendation that users write `ClassVar[ReadOnly[...]]` to avoid this?

Reply:
Sure.
I'd assume the wording of
> We recommend that in such circumstances `ReadOnly` is placed as the inner type qualifier, like `ClassVar[ReadOnly[...]]`

Is it ok to use "we" (the authors) within the PEP? Or perhaps "It is recommended ..." would be better?

Reply:
"Authors should prefer `ClassVar[ReadOnly[...]]` over `ReadOnly[ClassVar[...]]`..." ?

Reply:
I'd write something like "To avoid issues with runtime introspection, use `ClassVar[ReadOnly[...]]` instead of `ReadOnly[ClassVar[...]]`.

Reply:
I assumed the recommendation should apply in the general case 🤔
"To avoid issues with introspection, prefer nesting `ReadOnly` within other type qualifiers - use `ClassVar[ReadOnly[...]]` instead of `ReadOnly[ClassVar[...]]`."
Though it might give the wrong idea that you're always required to nest the type, or that one should put it within a generic that is not a type qualifier